File: vendor-web/src/types/project.ts
export type ProjectStatus = 'planning' | 'active' | 'on_hold' | 'completed' | 'cancelled'

export type ProjectPriority = 'low' | 'medium' | 'high' | 'urgent'

export type TaskStatus = 'todo' | 'in_progress' | 'review' | 'done'

export type TaskPriority = 'low' | 'medium' | 'high' | 'urgent'

export interface ProjectMilestone {
  id?: string
  title: string
  due_date?: string | null
  completed?: boolean
  completed_at?: string | null
}

export interface ProjectChecklistItem {
  id?: string
  text: string
  done?: boolean
}

export interface LinkedTaskSummary {
  id: string
  title: string
  status: TaskStatus
}

export interface Project {
  id: string
  vendor_id: string
  store_id?: string | null
  sales_area_id?: string | null
  items?: ProjectItemInput[]
  project_number: string
  name: string
  description?: string | null
  status: ProjectStatus
  priority: ProjectPriority
  customer_id?: string | null
  customer_name?: string | null
  customer_email?: string | null
  customer_phone?: string | null
  owner_id?: string | null
  owner_name?: string | null
  start_date?: string | null
  end_date?: string | null
  due_date?: string | null
  budget?: number | null
  currency: string
  progress_percent: number
  color?: string | null
  tags?: string[]
  milestones?: ProjectMilestone[]
  created_at: string
  updated_at: string
  completed_at?: string | null
  task_count?: number
  done_task_count?: number
}

export interface ProjectTask {
  id: string
  vendor_id: string
  project_id: string
  title: string
  description?: string | null
  status: TaskStatus
  priority: TaskPriority
  assignee_id?: string | null
  assignee_name?: string | null
  parent_task_id?: string | null
  parent_title?: string | null
  linked_task_ids?: string[]
  linked_tasks?: LinkedTaskSummary[]
  subtask_count?: number
  due_date?: string | null
  position: number
  labels?: string[]
  checklist?: ProjectChecklistItem[]
  created_at: string
  updated_at: string
  completed_at?: string | null
}

export interface ProjectOverview {
  total_projects: number
  by_status: Record<string, number>
  active_count: number
  overdue_count: number
  total_tasks: number
  open_tasks: number
  completed_tasks: number
  avg_progress: number
}

export interface ProjectListResponse {
  items: Project[]
  total: number
  page: number
  size: number
  pages: number
}

export interface ProjectItemInput {
  id: string
  name: string
  item_type: 'product' | 'service'
  sku?: string
  price?: number
}

export interface ProjectCreateInput {
  name: string
  description?: string
  status?: ProjectStatus
  priority?: ProjectPriority
  store_id?: string
  sales_area_id?: string
  items?: ProjectItemInput[]
  customer_id?: string
  customer_name?: string
  owner_id?: string
  owner_name?: string
  start_date?: string
  end_date?: string
  due_date?: string
}

export interface ProjectUpdateInput {
  name?: string
  description?: string
  status?: ProjectStatus
  priority?: ProjectPriority
  customer_id?: string | null
  customer_name?: string | null
  owner_id?: string | null
  owner_name?: string | null
  start_date?: string
  end_date?: string
  due_date?: string
  progress_percent?: number
  milestones?: ProjectMilestone[]
}

export interface ProjectTaskCreateInput {
  title: string
  description?: string
  status?: TaskStatus
  priority?: TaskPriority
  assignee_id?: string
  assignee_name?: string
  parent_task_id?: string
  linked_task_ids?: string[]
  due_date?: string
  position?: number
}

export interface ProjectTaskUpdateInput {
  title?: string
  description?: string
  status?: TaskStatus
  priority?: TaskPriority
  assignee_id?: string | null
  assignee_name?: string | null
  parent_task_id?: string | null
  linked_task_ids?: string[]
  due_date?: string
  position?: number
}

export interface TaskReorderItem {
  id: string
  status: TaskStatus
  position: number
}

export const PROJECT_STATUS_LABELS: Record<ProjectStatus, string> = {
  planning: 'Planning',
  active: 'Active',
  on_hold: 'On Hold',
  completed: 'Completed',
  cancelled: 'Cancelled',
}

export const PROJECT_PRIORITY_LABELS: Record<ProjectPriority, string> = {
  low: 'Low',
  medium: 'Medium',
  high: 'High',
  urgent: 'Urgent',
}

export const TASK_STATUS_LABELS: Record<TaskStatus, string> = {
  todo: 'To Do',
  in_progress: 'In Progress',
  review: 'Review',
  done: 'Done',
}

export const TASK_PRIORITY_LABELS: Record<TaskPriority, string> = {
  low: 'Low',
  medium: 'Medium',
  high: 'High',
  urgent: 'Urgent',
}

export const TASK_COLUMNS: TaskStatus[] = ['todo', 'in_progress', 'review', 'done']

File: backend/app/schemas/project.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from enum import Enum
from typing import Any, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


ORM = ConfigDict(from_attributes=True)


class ProjectStatus(str, Enum):
    PLANNING = "planning"
    ACTIVE = "active"
    ON_HOLD = "on_hold"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class ProjectPriority(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    URGENT = "urgent"


class TaskStatus(str, Enum):
    TODO = "todo"
    IN_PROGRESS = "in_progress"
    REVIEW = "review"
    DONE = "done"


class TaskPriority(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    URGENT = "urgent"


class MilestoneItem(BaseModel):
    id: Optional[str] = None
    title: str = Field(..., min_length=1, max_length=255)
    due_date: Optional[date] = None
    completed: bool = False
    completed_at: Optional[datetime] = None


class ChecklistItem(BaseModel):
    id: Optional[str] = None
    text: str = Field(..., min_length=1, max_length=500)
    done: bool = False


class ProjectItem(BaseModel):
    id: str
    name: str
    item_type: str = "product"  # product | service
    sku: Optional[str] = None
    price: Optional[float] = None


class ProjectBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    status: ProjectStatus = ProjectStatus.PLANNING
    priority: ProjectPriority = ProjectPriority.MEDIUM
    store_id: Optional[UUID] = None
    sales_area_id: Optional[UUID] = None
    customer_id: Optional[UUID] = None
    customer_name: Optional[str] = None
    owner_id: Optional[UUID] = None
    owner_name: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    due_date: Optional[date] = None
    budget: Optional[Decimal] = None
    currency: str = "INR"
    color: Optional[str] = Field(None, max_length=7)
    tags: Optional[list[str]] = None
    milestones: Optional[list[MilestoneItem]] = None
    items: Optional[list[ProjectItem]] = None


class ProjectCreate(ProjectBase):
    pass


class ProjectUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    status: Optional[ProjectStatus] = None
    priority: Optional[ProjectPriority] = None
    store_id: Optional[UUID] = None
    sales_area_id: Optional[UUID] = None
    items: Optional[list[ProjectItem]] = None
    customer_id: Optional[UUID] = None
    customer_name: Optional[str] = None
    owner_id: Optional[UUID] = None
    owner_name: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    due_date: Optional[date] = None
    budget: Optional[Decimal] = None
    currency: Optional[str] = None
    progress_percent: Optional[int] = Field(None, ge=0, le=100)
    color: Optional[str] = Field(None, max_length=7)
    tags: Optional[list[str]] = None
    milestones: Optional[list[MilestoneItem]] = None


class LinkedTaskSummary(BaseModel):
    id: UUID
    title: str
    status: str


class ProjectResponse(ProjectBase):
    model_config = ORM
    id: UUID
    vendor_id: UUID
    project_number: str
    progress_percent: int = 0
    customer_email: Optional[str] = None
    customer_phone: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    completed_at: Optional[datetime] = None
    task_count: Optional[int] = None
    done_task_count: Optional[int] = None


class ProjectListResponse(BaseModel):
    items: list[ProjectResponse]
    total: int
    page: int
    size: int
    pages: int


class ProjectOverviewResponse(BaseModel):
    total_projects: int = 0
    by_status: dict[str, int] = {}
    active_count: int = 0
    overdue_count: int = 0
    total_tasks: int = 0
    open_tasks: int = 0
    completed_tasks: int = 0
    avg_progress: float = 0.0


class TaskBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    status: TaskStatus = TaskStatus.TODO
    priority: TaskPriority = TaskPriority.MEDIUM
    assignee_id: Optional[UUID] = None
    assignee_name: Optional[str] = None
    parent_task_id: Optional[UUID] = None
    linked_task_ids: Optional[list[UUID]] = None
    due_date: Optional[date] = None
    position: int = 0
    labels: Optional[list[str]] = None
    checklist: Optional[list[ChecklistItem]] = None


class TaskCreate(TaskBase):
    pass


class TaskUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    status: Optional[TaskStatus] = None
    priority: Optional[TaskPriority] = None
    assignee_id: Optional[UUID] = None
    assignee_name: Optional[str] = None
    parent_task_id: Optional[UUID] = None
    linked_task_ids: Optional[list[UUID]] = None
    due_date: Optional[date] = None
    position: Optional[int] = None
    labels: Optional[list[str]] = None
    checklist: Optional[list[ChecklistItem]] = None


class TaskResponse(TaskBase):
    model_config = ORM
    id: UUID
    vendor_id: UUID
    project_id: UUID
    parent_title: Optional[str] = None
    linked_tasks: Optional[list[LinkedTaskSummary]] = None
    subtask_count: int = 0
    created_at: datetime
    updated_at: datetime
    completed_at: Optional[datetime] = None


class TaskReorderItem(BaseModel):
    id: UUID
    status: TaskStatus
    position: int


class TaskReorderRequest(BaseModel):
    items: list[TaskReorderItem]
